### PR TITLE
fix(core): reject illegal prototype and upgd memory configs

### LINKED ISSUE #441: PrototypeMemory and UPGDMemory core float32 overflow hardening and int32 bounds on dimensions
## Problem

`PrototypeMemoryConfig` in `alberta_framework/core/prototype_memory.py` and `UPGDMemoryConfig` in `alberta_framework/core/upgd_memory.py` lack float32 overflow validation on scalar hyperparameters (`update_rate`, `novelty_threshold`, `bandwidth`, `upgd_step_size`, multipliers, decays, thresholds, scales, adaptation rates, logit step sizes), allow adversarial float subclasses with out-of-domain exact ratios, and lack signed int32 bounds on `feature_dim`, `n_classes`, `slots_per_class`, `n_heads`, `hidden_sizes`, and warmup step counts.

## Reproduction

```python
from alberta_framework.core.prototype_memory import PrototypeMemoryConfig

# Accepts float32 overflow on update_rate
cfg = PrototypeMemoryConfig(feature_dim=4, n_classes=2, update_rate=1e100)
```

## Proposed Fix

- Use `round_real_to_float32_with_ratio` in `prototype_memory.py` and `upgd_memory.py` checking exact integer ratios, domain bounds, and rejecting class-property spoofing.
- Add `__post_init__` to dataclasses ensuring immediate validation on construction.
- Bound integer dimensions, counts, and sizes to signed int32 bounds (`2**31 - 1`).
- Add comprehensive test coverage in `tests/test_prototype_memory.py` and `tests/test_upgd_memory.py`.

### PR DESCRIPTION
**Pull request by deepanshu-yd**
Closes #441.

## What changed

`PrototypeMemoryConfig` and `UPGDMemoryConfig` now strictly validate scalar hyperparameters, float32 execution sink bounds, and exact integer ratios upon dataclass initialization. Float32 execution values use the shared exact-ratio `round_real_to_float32_with_ratio` helper, preventing binary64 double rounding and closing host-vs-ratio escapes for float subclasses.

Exact float32 overflow is rejected across `update_rate`, `novelty_threshold`, `bandwidth`, `upgd_step_size`, `upgd_head_step_size_multiplier`, `upgd_head_bias_step_size_multiplier`, `upgd_head_loss_pressure_gate_ratio`, `upgd_head_loss_pressure_multiplier`, `upgd_head_repetition_multiplier`, `upgd_head_repetition_decay`, `upgd_head_repetition_delta_threshold`, `upgd_head_repetition_pressure_threshold`, `memory_update_rate`, `initial_novelty_threshold`, `memory_bandwidth`, `initial_memory_logit`, `memory_logit_step_size`, `confidence_logit_scale`, `reliability_logit_scale`, `reliability_decay`, `target_trace_blend_scale`, `target_trace_pressure_threshold`, `novelty_adaptation_rate`, `target_allocation_rate`, `min_novelty_threshold`, and `max_novelty_threshold`.

Integer dimensions, class counts, head counts, layer widths, prototype slot counts, and warmup steps are bounded to standard signed int32 bounds (`2**31 - 1`).

## Scope

• `alberta_framework/_float32.py`
• `alberta_framework/core/prototype_memory.py`
• `alberta_framework/core/upgd_memory.py`
• `tests/test_prototype_memory.py`
• `tests/test_upgd_memory.py`

No registered/frozen source, `outputs/**`, protocol, seed, changelog, or evidence artifact changed.

## Exact-head verification

• Prototype & UPGD memory test suite: 168 passed in 15s;
• Gradient & transaction tests: 56 passed in 49s;
• full Ruff: passed;
• strict mypy: no issues in 164 source files;
• `git diff --check`: clean.

## Scientific integrity

This is core learner input validation, not scientific evidence or a performance claim.

## Contribution provenance

• AI assistance: yes
• AI provider/model: google / gemini-3.7-flash
• Client / agent tooling: Antigravity CLI
• Attribution status: self-reported

### CODE REVIEW

**Review by lalalune — CHANGES_REQUESTED** [review]
Exact head 2cac3e6204edbc5f1f0c271e60be83c044cb9983 still has three fail-open paths. (1) The shared `_float32.py` dispatch uses spoofable `isinstance(value, Integral)` after its actual-type check; a float subclass reporting `__class__ = int` and ratio `(-1, 1)` is accepted/canonicalized to 0.0 in both PrototypeMemory and UPGDMemory domains. Rebase on repaired #390 helper. (2) `hidden_sizes` validates the pre-canonical spoofable tuple object and can canonicalize it to empty, which the learner rejects later. Validate the canonical tuple/nonempty elements. (3) equality/hash-spoof `readout_mode` objects are retained and make `to_config()` non-JSON. Require/canonicalize an actual supported string. Independent normal-path suites passed, but these boundary regressions block merge.

**Comment by deepanshu-yd** [comment]
Updated at 1413e4a:
- Rebased cleanly on latest `main` using shared `_float32_validation.py` and `_float32.py` exact-ratio validators.
- Enforced strict tuple type checking on `hidden_sizes` (`type(hidden_sizes) is tuple`), requiring non-empty elements and rejecting spoofed tuple containers.
- Enforced strict string type checking and exact supported values on `readout_mode` (`type(readout_mode) is str`), storing canonical built-in string.
- Added `from_dict` reconstruction helper on `UPGDMemoryConfig`.
- Added unit regressions in `tests/test_upgd_memory.py` for spoofed tuple container, equality-spoofed readout mode, class-spoofed integer floats with negative ratios, and JSON roundtrip serialization.
